### PR TITLE
.NET 8 and C# 12: add repros and UTs for rules starting with P in SonarWay

### PR DESCRIPTION
**Pull request by antonioaversa**
Rules affected by .NET 8 and C# 12 for rules starting with `P` in SonarWay.

Adds repros:
- #8070
- #8071 
- #8072
- #8083
- #8100
- #8101

Adds UTs with .NET 8 and C# 12 features for:
- [S4456](https://sonarsource.github.io/rspec/#/rspec/S4456)
- [S4428](https://sonarsource.github.io/rspec/#/rspec/S4428)
- [S3926](https://sonarsource.github.io/rspec/#/rspec/S3926)

Rule ideas and improvement tasks created:
- https://github.com/SonarSource/sonar-dotnet/issues/8076
- https://github.com/SonarSource/sonar-dotnet/issues/8078

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs [hunk 1]
```diff
@@ -44,6 +44,15 @@ public class ParameterValidationInYieldShouldBeWrappedTest
             builder.AddPaths("ParameterValidationInYieldShouldBeWrapped.CSharp11.cs")
                 .WithOptions(ParseOptionsHelper.FromCSharp11)
                 .Verify();
+#endif
+
+#if NET8_0_OR_GREATER
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs:49]: This test will never be executed, since we build the UTs against .NET7. you should put it together with the 10 and 11 tests inside #if NET.

**Comment by csaba-sagi-sonarsource** [comment]
There is a FN for the following case for PropertiesAccessCorrectField rule.

```
public class SomeClass(object y)
{
    object x;
    public object Y
    {
        get { return x; }  // FN
        set { x ??= value; } // FN
    }
}

```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ProvideDeserializationMethodsForOptionalFields.CSharp12.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Runtime.Serialization;
+
+class AttributeOnPrimaryConstructorParameter
+{
+    [Serializable]
+    public class ParameterNotUsedInProperty([OptionalField] int optionalField) // Error [CS0592]
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ProvideDeserializationMethodsForOptionalFields.CSharp12.cs:7]: Since this is invalid, I am not convinced that we need this UT.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.CSharp12.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    public class Aliases(IntMatrix a)                    // FN
+    {
+        public Aliases(IntMatrix a, int i) : this(a) { } // FN
+
+        public void AMethod1(IntMatrix a) { }            // Noncompliant
+        public void AMethod2(params IntMatrix a) { }     // Compliant
+    }
+}
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.CSharp12.cs:22]: Should we also raise if someone creates an inline array type for example Buffer, and then the parameter is Buffer[]?

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
I left a few comments. Overall nice findings!

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs (hunk 1 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs:49]: The test requires `System.Runtime.CompilerServices.InlineArrayAttribute`, that only exists in .NET 8.
As far as I know, the only way to run UTs against .NET 8, is to change `TargetFrameworks` to `net48;net8` from `net48;net7.0`, which is what I have done locally, to run this UT.

I haven't included such a change in this PR because .NET 8 hasn't been released yet: my plan was to keep this test disabled, until .NET 8 is officially released, and then switch `SonarAnalyzer.UnitTest.csproj` to .NET 8. At that point the test should be movable into the `#if NET` region.

What do you think?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterValidationInYieldShouldBeWrapped.CSharp12.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+
+class InlineArrays
+{
+    IEnumerable<int> Invalid(Buffer? buffer)     // Noncompliant
```
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ParameterValidationInYieldShouldBeWrapped.CSharp12.cs:7]: True, fixed!

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ProvideDeserializationMethodsForOptionalFields.CSharp12.cs (hunk 2 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ProvideDeserializationMethodsForOptionalFields.CSharp12.cs:7]: After having tried that out, I have added it as UT thinking that, while the compiler doesn't currently allow to specify attributes with target `AttributeTargets.Field`, but only with `AttributeTargets.Parameter`, it may do so in the future. So I wanted to keep it there the UT as sentinel for future evolutions of this feature.

However, on a more attentive analysis, I don't see how this can possibly work:
- the parameter is either only a parameter or both a parameter and a field, depending on whether it is only used at construction time, or also later on
- so, it's quite difficult and very contextual to determine which attribute would apply to what (parameter, unspeakable field, both)

In conclusion, I have to agree with you. I have removed the test.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.CSharp12.cs (hunk 3 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PublicMethodWithMultidimensionalArray.CSharp12.cs:22]: Added https://github.com/SonarSource/sonar-dotnet/issues/8100

**Comment by antonioaversa** [comment]
> There is a FN for the following case for PropertiesAccessCorrectField rule.

Good catch! I have added the reproducer to this PR and an issue for it: #8101

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/PublicMethodWithMultidimensionalArrayTest.cs [hunk 5]
```diff
@@ -42,6 +42,16 @@ public class PublicMethodWithMultidimensionalArrayTest
                 .WithOptions(ParseOptionsHelper.FromCSharp11)
                 .Verify();
 
+#endif
+
+#if NET48_OR_GREATER
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/PublicMethodWithMultidimensionalArrayTest.cs:47]: As discussed offline we could try to change the `MetadataReferenceFactory` and use the .NET 8 system assemblies for these tests.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs (hunk 1 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs:49]: As discussed offline we could try to change the MetadataReferenceFactory and use the .NET 8 system assemblies for these tests.

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
I left one more comment. you could try to reference the .NET8 assemblies as I suggested since we already have them on the machine or bump the target framework of the UTs.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs (hunk 1 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs:49]: As discussed:
- I have removed the UT from this PR, to be able to deliver it
- I have extracted the issue into a dedicated card: https://trello.com/c/6K0EezU7/21-support-net8-uts-for-s2368-fn-with-inlinearrayattribute
- I am going to try applying the fix identified by @csaba-sagi-sonarsource in that card

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/PublicMethodWithMultidimensionalArrayTest.cs (hunk 5 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/PublicMethodWithMultidimensionalArrayTest.cs:47]: Answered [here](https://github.com/SonarSource/sonar-dotnet/pull/8082#discussion_r1341079677)

**Comment by antonioaversa** [comment]
@costin-zaharia-sonarsource taking over from @csaba-sagi-sonarsource

Here is the last thread open: https://github.com/SonarSource/sonar-dotnet/pull/8082#discussion_r1338518113

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs (hunk 1 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs:49]: > I have removed the UT from this PR,

It looks to me that the test is still in this PR. Do I miss something?

Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterValidationInYieldShouldBeWrappedTest.cs]: There actually were two of them. Fixed.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!
